jupyterlab/handlers/yjs_echo_ws.py

```diff
@@ -25,170 +32,170 @@
-            for _, (_, _, client) in room.clients.items():
-                client.room_id = new_room_id
-            ROOMS.pop(room_id)
-            ROOMS[new_room_id] = room
+        self._message_queue.put_nowait(message)
+        if message[0] == RENAME_SESSION:
+            # The client moved the document to a different location. After receiving this message, we make the current document available under a different url.
+            # The other clients are automatically notified of this change because the path is shared through the Yjs document as well.
+            new_room_name = message[1:].decode("utf-8").split(":", 1)[1]
+            WEBSOCKET_SERVER.rename_room(new_room_name, from_room=self.room)
             # send rename acknowledge
-            self.write_message(bytes([ServerMessageType.RENAME_SESSION, 1]), binary=True)
-            # print("renamed room to " + new_room_id + ". Old room name was " + room_id)
-        elif room:
-            if message[0] == 0:  # sync message
-                read_sync_message(self, room.ydoc.ydoc, message[1:])
-            for client_id, (loop, hook_send_message, _) in room.clients.items():
-                if self.id != client_id:
-                    loop.add_callback(hook_send_message, message)
-
-    def on_close(self):
-        # print("[YJSEchoWS]: close")
-        room = ROOMS.get(self.room_id)
-        room.clients.pop(self.id)
-        if not room.clients:
-            ROOMS.pop(self.room_id)
-            # print("[YJSEchoWS]: close room " + self.room_id)
-
+            self.write_message(bytes([RENAME_SESSION, 1]), binary=True)
+
+    def on_close(self) -> bool:
+        # stop serving this client
+        self._message_queue.put_nowait(b"")
+        if not self.room.clients:
+            # keep the document for a while after every client disconnects
+            self.room.cleaner = asyncio.create_task(self.clean_room())
         return True
 
-    def check_origin(self, origin):
-        # print("[YJSEchoWS]: check origin")
+    async def clean_room(self) -> None:
+        await asyncio.sleep(60)
+        self.room.watcher.cancel()
+        self.room.document.unobserve()
+        WEBSOCKET_SERVER.delete_room(room=self.room)
+
+    def maybe_save_document(self, event):
+        # unobserve and observe again because the structure of the document may have changed
+        # e.g. a new cell added to a notebook
+        self.room.document.unobserve()
+        self.room.document.observe(self.maybe_save_document)
+        if self.saving_document is not None and not self.saving_document.done():
+            # the document is being saved, cancel that
+            self.saving_document.cancel()
+            self.saving_document = None
+        self.saving_document = asyncio.create_task(self.save_document())
+
+    async def save_document(self):
+        # save after 1 second of inactivity to prevent too frequent saving
+        await asyncio.sleep(1)
```

Comment:
I would move the rate limiter in `maybe_save_document`. So the `save_document` method does what it says and nothing else.

Reply:
:+1:

Reply:
The problem is that the `save_document` task can be cancelled (if a new document change happens less than one second later), so this won't work.
I'll just rename `maybe_save_document` to `on_document_change`, and `save_document` to `maybe_save_document`. Feedback welcome!